pytorch_lightning/loops/epoch/training_epoch_loop.py

```diff
@@ -268,65 +265,76 @@ def _should_accumulate(self) -> bool:
+        swapped = [[] for _ in range(n_opt)]
+        for optimizers in tbptt_outputs:
+            for opt_idx, values in optimizers.items():
+                swapped[opt_idx].append(values)
+        return swapped
 
-        # track the outputs to reduce at the end of the epoch
-        for opt_idx, opt_outputs in enumerate(batch_end_outputs):
-            # with 1 step (no tbptt) don't use a sequence at epoch end
-            if isinstance(opt_outputs, list) and len(opt_outputs) == 1:
-                opt_outputs = opt_outputs[0]
+    @staticmethod
+    def _prepare_outputs_training_batch_end(
+        batch_output: _BATCH_OUTPUTS_TYPE,
+    ) -> Union[List[List[Dict[str, Any]]], List[Dict[str, Any]]]:
+        """Processes the outputs from the batch loop into the format passed to the ``training_batch_end`` hook.
 
-            epoch_output[opt_idx].append(opt_outputs)
+        ``(tbptt_steps, n_opt) -> (n_opt, tbptt_steps)``. The optimizer dimension might have been squeezed.
+        """
+        n_opt = max(max(tbptt.keys()) for tbptt in batch_output) + 1
+        swapped = TrainingEpochLoop._swap_optimizers_and_tbptt(batch_output, n_opt)
+        # squeeze optimizer dimension
+        return swapped[0] if len(swapped) == 1 else swapped
 
     @staticmethod
-    def _prepare_outputs(
-        outputs: List[List[List[OutputResult]]], batch_mode: bool
-    ) -> Union[List[List[List[Dict]]], List[List[Dict]], List[Dict], Dict]:
-        """Extract required information from batch or epoch end results.
-
-        Args:
-            outputs: A 3-dimensional list of ``OutputResult`` objects with dimensions:
-                ``[optimizer outs][batch outs][tbptt steps]``.
+    def _prepare_outputs_training_epoch_end(
+        batch_outputs: _OUTPUTS_TYPE,
+        automatic: bool,
+    ) -> Union[List[List[List[Dict[str, Any]]]], List[List[Dict[str, Any]]], List[Dict[str, Any]], Dict[str, Any]]:
+        """Processes the outputs from the batch loop into the format passed to the ``training_epoch_end`` hook.
 
-            batch_mode: If True, ignore the batch output dimension.
+        ``(n_batches, tbptt_steps, n_opt) -> (n_opt, n_batches, tbptt_steps)``.
+        All single-element dimensions might have been squeezed.
 
-        Returns:
-            The cleaned outputs with ``OutputResult`` objects converted to dictionaries.
-            All list dimensions of size one will be collapsed.
+        This processing is necessary because the format of the inputs to the ``training_epoch_end`` hook does not
+        match the loop structure and because empty dimensions are squeezed. This could break with loop customization.
         """
-        processed_outputs = []
-        for opt_outputs in outputs:
-            # handle an edge case where an optimizer output is the empty list
-            if len(opt_outputs) == 0:
+        # `batch_outputs` (plural) is the same as `epoch_end_output` (singular)
+        if not batch_outputs:
+            return {}
```

Comment:
we return a dict here but I don't remember why. the type hint says we can but I don't think that's correct.